### ISSUE TITLE
Add MathMLView Control

### ISSUE THREAD

**Issue by insinfo** [Microsoft/xaml-standard#192]
- **Overview** 
_The proposal of the MathMLView is rendering mathematical formulae using the LaTeX typsetting style or MathML documents. 
It provides an interactive view that can be used for browsing and editing Math formulae ._
- **Scenario** 
The idea here and make life easier for the developer, as can be seen in both the Play Store and the Apple Store, to a significant amount of Apps on the theme math that depend on rendering math formulas, this extends also to the desktop. Mathematics is everywhere. How do you have this one here too?
![mathml view](https://cloud.githubusercontent.com/assets/12227024/26763471/b5bbf614-4929-11e7-9b83-7c964cef840d.png)

- **Example in XAML.**
```xaml
<MathMLView  Formula="\left(x^2 + 2 \cdot x + 2\right) = 0" />
```
- **Links** 
https://play.google.com/store/search?q=Math&c=apps&hl=pt_BR
http://helm.cs.unibo.it/mml-widget/
https://github.com/ForNeVeR/wpf-math
https://pt.wikipedia.org/wiki/MathML

**Comment by birbilis** [comment]
Obviously needs to support binding inside the formula string too, probably with some escape start/end syntax if latex/MathML has such

**Comment by insinfo** [comment]
@birbilis 
yes

**Comment by dotMorten** [comment]
IMHO this is so specialized and should stay out of the common basic set and be offered as a 3rd party control or extension.

**Comment by Duranom** [comment]
I agree with @dotMarten that this an extremely specialized concept which use can be found only in specialized applications, not to mention you would have to support many variant of notations.

**Comment by insinfo** [comment]
@dotMorten 
I do not agree with you, a robust and complete markup language standard to develop graphical inteface can not be limited to buttons ...
For example the web standards, it has support for MathML

**Comment by crutkas** [comment]
@insinfo while yes it is in the HTML5 spec, I believe only Firefox supports it.  Also quickly glancing, it does not seem like either iOS or Android also support this inbox.

I think it is a great idea.  How would you stack rank this control in terms of other controls you've been asking for?

**Comment by birbilis** [comment]
> SVG backend, which renders MathML into SVG.
http://helm.cs.unibo.it/mml-widget/

Wonder if Edge team has MathML or if in the works, if they made a renderer to SVG. From that one could get XAML Paths I guess (there's issue logged here for SVG support btw)
Some related discussion is at:
https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/6508572-mathml

Wikipedia article on MarhML says Chrome has promised MathML too and has some example XML syntax for it and comparison to other formats. Among those I'd prefer AsciiMath, not sure how flexible it is though.

Btw, latex syntax is different beast, MathML can just contain it in an annotation but it's not to be parsed from there (see example at wikipedia)

**Comment by MathiasMagnus** [comment]
While I do agree with the initiative, I do have a few questions/suggestions:

- Why do you insist on giving a textual shorthand for the formulas, when the whole purpose of MathML is to give it XML syntax. Why cannot that be embedded into XAML?
```XML
<MathMLView Property1="" Property2="">
  <apply>
    <sin/>
      <ci>x</ci>
  </apply>
<MathMLView/>
```
- I think that agreeing on this functionality as part of XAML standard is easier if the language (LaTeX, asciiMath, etc.) of the shorthand is not included in the earliest versiion. Doing so we can postpone such discussions once implementations are in place.
- As a C++ developer, it has struck me gazillions of times how lacking the STL is. (XML, JSON, still missing, Filesystem still not implemented by as a whole, networking still not quite done, graphics, etc.) The most compelling thing about .NET is that you can go roughly all the way without having to resort to 3rd party libraries. Making XAML Standard a swiss army knife would only promote its usage as opposed to sticking to its various dialects.
  - I'd say go bold and put as much awesome into XAML Standard as possible.
- Since I'm a physics PhD and I find it a joke how lacking MathML support is in 2017 on the browser side, so I'm all for any attempt to pushing it into places where it makes sense.

**Comment by insinfo** [comment]
I totally agree with you, @MathiasMagnus

> I'd say go bold and put as much awesome into XAML Standard as possible.

 I think it's an excellent suggestion for your idea.

**Comment by insinfo** [comment]
I added your suggestion to my initial proposal

**Comment by insinfo** [comment]
I believe in the idea that there are already many standards and a lot of segmentation, what is lacking, in my opinion is a robust and complete standard, enough to reinvest the wheel with each new project. We need a universal UI standard that is robust and complete in all areas of software development, whether in physics, math, entertainment, business, education, government, health, etc. Because with a robust base and complete graphic interface controls it is possible to embrace all these areas, which needs software with fast, beautiful, fluid and intuitive graphic interface. A standard that can be used in both Desktop, Mobile, WEB, Embedded Devices, self-service machines etc. With the introduction of WebAssembly this is made easier, implementing a Universal UI Framework in XAML.

**Comment by Duranom** [comment]
A thing that worries me is that should there be a new version of the MathML standard, when and how to have it updated in the xaml standard. It is not unknown that there have been, and still are, standards in the framework(s) that have not been updated and now outdated in such a way that in almost all new cases should be avoided, for example like XSLT. For MathML that would be really undesirable seeing as the fields it is used in.

**Comment by birbilis** [comment]
When something is mature enough it can be trusted to not change all the time

**Comment by Happypig375** [comment]
Why MathML? Why not the more standard LaTeX format?

**Comment by MathiasMagnus** [comment]
Three reasons:

1. Well, because MathML is far easier to implement. It's format is XML, so the parser is done.
2. It would be 100% portable with HTML content. Though LaTeX is portable with LaTeX documents, but I think the audience of that is far smaller than the web in itself.
3. You say LaTeX format, I say [Unicode linear format](http://www.unicode.org/notes/tn28/UTN28-PlainTextMath-v3.pdf) (found in MS Word), others say something else...

MathML is a W3C standard and deserves better.

**Comment by Happypig375** [comment]
Four opposing arguments:
1. https://tex.stackexchange.com/questions/57717/relationship-between-mathml-and-tex

>MathML is very verbose and is a pain to write manually, which is no problem if your code is automatically generated. For programs it is easy to parse, since it is XML.


>TeX math code was designed to be written by humans and results in much shorter code. I cannot answer your question if one format will replace the other. As I mentioned they serve different purposes.


Sure, MathML is easy for computers to parse. What about humans? Can humans parse MathML effeciently? I cannot. Instead, LaTeX looks more natural and I can understand what a piece of LaTeX means, far better than MathML. Using MathML instead of LaTeX just increases maintenance costs. 

2. https://developer.mozilla.org/en-US/docs/Web/MathML/Authoring

>There are many simple notations (e.g. wiki or markdown syntaxes) to generate HTML pages. The same is true for MathML: for example ASCII syntaxes as used in calculators or the more powerful LaTeX language, very popular among the scientific community.


Even Mozilla describes LaTeX as being very popular and it is used to generate MathML. Therefore, I don't agree with you saying that MathML is far more popular than LaTeX - People don't write MathML, they generate them using LaTeX or other similar formats. 

3. https://support.office.com/en-us/article/Linear-format-equations-using-UnicodeMath-and-LaTeX-in-Word-2e00618d-b1fd-49d8-8cb4-8d17f25754f8

>Linear format equations using UnicodeMath and LaTeX in Word


Besides UnicodeMath, LaTeX is supported in Word too. Also, MathML is not there.

A quick Google search shows that there is a LaTeX renderer for WPF , but no relevant material shows up for the search on UnicodeMath , so the difference in popularity is so obvious. Therefore, I picked LaTeX over other formats.

4. Just because MathML is a W3C standard does not mean that it should deserve better. https://www.google.com.hk/search?hl=en-GB&ie=UTF-8&source=android-browser&q=failed+w3c+standarda&gfe_rd=cr&dcr=0&ei=EvknWq_kMsPW4gLS_aqQDA

>W3C's failed Do Not Track crusade tumbles to ad-blockers' Vietnam


Here is an example of a failed W3C standard. So, we should not determine what a format deserves just based on whether it is a standard or not.

Any rebuttals?

**Comment by insinfo** [comment]
My sugestion
This control can be implemented more generally, and an interface is provided so that it can easily be extended to other mathematical languages. Initially, it would come with a basic implementation for Latex and MathML

I updated my post, to leave the idea of the most generic control.

**Comment by Happypig375** [comment]
Hi everyone, there is an implementation of this right now for Xamarin.iOS:
https://github.com/verybadcat/CSharpMath

Although only Xamarin.iOS is supported right now, there is a shared front end which can be extended to support other platforms.
Can anybody here take a look at this and give some feedback?
@verybadcat would like some help.

Although only LaTeX is supported now, a MathML-to-LaTeX converter in the library could be added in the future. With this library, we won't need to wait Microsoft to implement this.

What do you think?

**Comment by insinfo** [comment]
@Happypig375 
Thanks for looking at this control, I believe this control is very important for a robust and complete XAML Standard, and for modern next-generation software that involves visual programming, interactivity with neural networks, and more.
I think your initiative is very good, we need more initiatives like this,
because it seems that if we are to wait for Microsoft that will not leave the paper.

**Comment by danvanderboom** [comment]
Starting with Android API 26, fonts can be loaded as resources in Xamarin Android projects.
https://developer.xamarin.com/guides/android/application_fundamentals/resources_in_android/part_7_-_fonts/
